klippy/extras/dockable_probe.py

```diff
@@ -0,0 +1,596 @@
+            try:
+                p = [None] * 3
+                val = config.get(name, None)
+                if not val:
+                    return None
+                p[0:c] = map(lambda x: float(x.strip()),
+                             val.split(','))[0:c]
+            except:
+                e = "Unable to parse {0} in {1}"
+                raise config.error(e.format(self.name, name))
+            return p
+
+        self.approach_position = parseCoord('approach_position')
+        self.detach_position   = parseCoord('detach_position')
+        self.dock_position     = parseCoord('dock_position')
+        self.safe_z_position   = parseCoord('safe_z_position', 2)
+
+        self.dock_angle, self.approach_distance = self._get_vector(
+                                                        self.dock_position,
+                                                        self.approach_position)
+        self.detach_angle, self.detach_distance = self._get_vector(
+                                                        self.dock_position,
+                                                        self.detach_position)
+
+        #Pins
+        ppins = self.printer.lookup_object('pins')
+        pin = config.get('pin')
+        pin_params = ppins.lookup_pin(pin, can_invert=True, can_pullup=True)
+        mcu = pin_params['chip']
+        mcu.register_config_callback(self._build_config)
+        self.mcu_endstop = mcu.setup_pin('endstop', pin_params)
+
+        # Wrappers
+        self.get_mcu              = self.mcu_endstop.get_mcu
+        self.add_stepper          = self.mcu_endstop.add_stepper
+        self.get_steppers         = self.mcu_endstop.get_steppers
+        self.home_wait            = self.mcu_endstop.home_wait
+        self.query_endstop        = self.mcu_endstop.query_endstop
+        self.finish_home_complete = self.wait_trigger_complete = None
+
+        pstate = ProbeState(config, self)
+        self.get_probe_state = pstate.get_probe_state
+
+        #Custom User Macros
+        gcode_macro = self.printer.load_object(config, 'gcode_macro')
+        self.pre_attach_gcode = gcode_macro.load_template(
+            config, 'pre_attach_gcode', '')
+        self.attach_gcode = gcode_macro.load_template(
+            config, 'attach_gcode', '')
+        self.post_attach_gcode = gcode_macro.load_template(
+            config, 'post_attach_gcode', '')
+        self.pre_detach_gcode = gcode_macro.load_template(
+            config, 'pre_detach_gcode', '')
+        self.detach_gcode = gcode_macro.load_template(
+            config, 'detach_gcode', '')
+        self.post_detach_gcode = gcode_macro.load_template(
+            config, 'post_detach_gcode', '')
+
+        #Gcode Commands
+        self.gcode.register_command('ATTACH_PROBE',
+                                    self.cmd_ATTACH_PROBE)
+        self.gcode.register_command('DETACH_PROBE',
+                                    self.cmd_DETACH_PROBE)
```

Comment:
add a desc=self.cmd_DETACH_PROBE_help otherwise the gcode will not show up in help. Could be as simple as "Check probe status and detach probe"